src/rmc-compiler/build.rs

```diff
@@ -0,0 +1,15 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use std::env;
+use std::path::PathBuf;
+
+/// Configure the compiler to build rmc-compiler binary. We currently support building
+/// rmc-compiler with nightly only. We also link to the rustup rustc_driver library for now.
+pub fn main() {
+    // Add rustup to the rpath in order to properly link with the correct rustc version.
+    let rustup_home = env::var("RUSTUP_HOME").unwrap();
+    let rustup_tc = env::var("RUSTUP_TOOLCHAIN").unwrap();
+    let lib_path = [&rustup_home, "toolchains", &rustup_tc, "lib"].iter().collect::<PathBuf>();
+    println!("cargo:rustc-link-arg-bin=rmc-compiler=-Wl,-rpath,{}", lib_path.to_string_lossy());
```

Comment:
Just checking in here, because I'll have to work on the install later, that you're dynamic linking against a `rustup` provided rust compiler library?

Is there documentation on how this works somewhere? (e.g. I'm used to rust's normal static linking, so dynamic linking to a rust library is not something I'm familiar with)

Also, this might complication installation later. For one, we need the correct relative RPATH here not an absolute path like it is now, but also we'll need to distribute the correct rustc version, right? Or is there a stable-ish ABI here?

If there's not documentation about this, I think we need to write one.

Reply:
Yes, you are correct. It is linking to a rustup shared library.  We will need to make it relative if we want to distribute the binaries or if we eventually install it via rustup.

This should work if the user wants to build / install it with cargo.

Reply:
Do you know where docs about this are? I searched "rust shared library rustc_driver" and I'm not finding anything helpful. I'm not sure how fragile this approach is. (But I sure like what it does for CI times! Just trying to understand the implications for installation later...)

Reply:
I couldn't find any doc either. I did a lot of reverse engineering in other tools that use rustc to figure this out and I also talked to a few folks.

This seems to be the de facto way to build tools on the top of rustc. See https://github.com/rust-lang/rust-clippy/blob/830f2205d4002b26f7fd36d8467fb491eb2306ac/src/driver.rs, https://github.com/rust-lang/miri/blob/035933186957cf81c488261fb48a98bf523e8006/src/lib.rs and this little example: https://doc.rust-lang.org/beta/unstable-book/language-features/plugin.html?highlight=rustc_private#lint-plugins